AdsLib/standalone/AmsRouter.cpp

```diff
@@ -51,22 +57,40 @@ long AmsRouter::AddRoute(AmsNetId ams, const std::string& host)
         }
     }
 
-    auto conn = connections.emplace(std::unique_ptr<AmsConnection>(new AmsConnection { *this, hostAddresses.get()}));
-    if (conn.second) {
-        /** in case no local AmsNetId was set previously, we derive one */
-        if (!localAddr) {
-            localAddr = AmsNetId {conn.first->get()->ownIp};
+    connection_attempts[ams] = {};
+    lock.unlock();
+
+    try {
+        auto new_connection = std::unique_ptr<AmsConnection>(new AmsConnection { *this, hostAddresses.get()});

```

Comment:
Hi @nicolatimeus,

first of all thanks for PR and sorry I didn't respond earlier, but I just came back from vacation. I wonder why you implement it with another "waitqueue". Wouldn't it be sufficient to do something like:
```
std::unique_lock<std::recursive_mutex> lock(mutex);
lock.lock();
const auto oldConnection = GetConnection(ams);
... 

// Give other threads the opportunity to use "connections", while we establish a new TCP connection
lock.unlock();
auto new_connection = std::unique_ptr<AmsConnection>(new AmsConnection { *this, hostAddresses.get()});
lock.lock();
auto conn = connections.emplace(std::move(new_connection));
...

```

Reply:
Hi, thanks for your reply. 
It is for adding protection for the case when a thread calls `AddRoute` while another one is already performing a connection attempt on the same route without holding the global lock. With this implementation basically we have a wait queue per route. Another approach for solving this would be allowing allowing all threads to initiate a new connection attempt and then all except one must drop their connection (in this case we have to define the behaviour for the case when the different threads call `AddRoute` with the same route but different ip address).

Reply:
Okay, since I think racing "AddRoutes(/DelRoutes)" should be rare I think we can try something simpler like

Splitting AddRoute() into two functions: GetExistingRoute(), AddNewRoute();
then we move the lock inside these functions. And get something like:

```
AddRoute() {
    // WARNING extrem pseudo code:
    if (GetExistingRoute()  == PORTALREADYINUSE or 0 {
        return 0;
    }
    AddNewRoute();
    return GetExistingRoute(); 
}

```

Reply:
This change was meant to modify the behavior of existing code as little as possible, however if it is acceptable to open new connections and immediately drop them in case of contention (that we assume will be unlikely) maybe we can also try something like this:

```c++
....
    lock.unlock();
    
    auto newConnection = std::unique_ptr<AmsConnection>(new AmsConnection { *this, hostAddresses.get()});
    
    lock.lock();

    // if another thread added a connection for this route concurrently, drop the new connection and try again
    if (mapping.find(ams) != mapping.end()) {
        lock.unlock();
        newConnection.reset();
        return AddRoute(ams, host);
    }

    auto conn = connections.emplace(std::move(newConnection));
...
```

Reply:
Hmm, well we still change the behavior a bit. Since we allow parallel connection attempts for different AmsNetIds, It can happen that you end up in the situation you describe. But I agree this should be rare and can easily avoided by the application, as the application could add a similar waitqueue for `host` in front of the `AddRoute()` call.

I guess parallel TCP connections at startup are a use case for applications connecting to different TwinCAT systems so I think your waitqueue is worth the extra effort. Now I think I took enough time to understand your code and my worry that AddRoute/DelRoute are still racing has gone. I am testing this change on our CI with just a minimal adjustment:

```
+   void AwaitConnectionAttempts(const AmsNetId& ams, std::unique_lock<std::recursive_mutex>& lock);
    void DeleteIfLastConnection(const AmsConnection* conn);
 -  void AwaitConnectionAttempts(const AmsNetId& ams, std::unique_lock<std::recursive_mutex>& lock);
```

Reply:
Additionally I ran uncrustify on your commit and added some background to the commit message. I would merge it https://github.com/Beckhoff/ADS/commit/5dbbfade98d4b8e93b623d8519d0e7c931a6c4d2  like this after the initial bug #208  is confirmed to be fixed by this.

Reply:
Thank you! are you aware of any other parts of the library that may perform time-consuming operations with the global lock held?

Reply:
No. Everything else runs on one thread per TCP connection and an additional handover for notifications. So there is nothing I am aware of.

Are you working together with @ashmookler1 or did you just ran into the same issue?

Reply:
@pbruenn I am working with @nicolatimeus, if he is satisfied with the fix then I am too. Thank you!

Reply:
merged as https://github.com/Beckhoff/ADS/commit/8e8fb2a9acff06cc617f351ea35dbb9e56f0cf97

Note: I dropped the extra includes, as they aren't necessary on any of our build tests.